core/chains/evm/gas/universal_estimator.go

```diff
@@ -0,0 +1,413 @@
+	return &UniversalEstimator{
+		client:   client,
+		logger:   logger.Named(lggr, "UniversalEstimator"),
+		config:   cfg,
+		chainID:  chainID,
+		l1Oracle: l1Oracle,
+		wg:       new(sync.WaitGroup),
+		stopCh:   make(chan struct{}),
+	}
+}
+
+func (u *UniversalEstimator) Start(context.Context) error {
+	return u.StartOnce("UniversalEstimator", func() error {
+		if u.config.BumpPercent < MinimumBumpPercentage {
+			return fmt.Errorf("BumpPercent: %s is less than minimum allowed percentage: %s",
+				strconv.FormatUint(uint64(u.config.BumpPercent), 10), strconv.Itoa(MinimumBumpPercentage))
+		}
+		if u.config.EIP1559 && u.config.RewardPercentile > ConnectivityPercentile {
+			return fmt.Errorf("RewardPercentile: %s is greater than maximum allowed percentile: %s",
+				strconv.FormatUint(uint64(u.config.RewardPercentile), 10), strconv.Itoa(ConnectivityPercentile))
+		}
+		if u.config.EIP1559 && u.config.BlockHistorySize == 0 {
+			return fmt.Errorf("BlockHistorySize is set to 0 and EIP1559 is enabled")
+		}
+		u.wg.Add(1)
+		go u.run()
+
+		return nil
+	})
+}
+
+func (u *UniversalEstimator) Close() error {
+	return u.StopOnce("UniversalEstimator", func() error {
+		close(u.stopCh)
+		u.wg.Wait()
+		return nil
+	})
+}
+
+func (u *UniversalEstimator) run() {
+	defer u.wg.Done()
+
+	t := services.NewTicker(u.config.CacheTimeout)
+	for {
+		select {
+		case <-u.stopCh:
+			return
+		case <-t.C:
+			if u.config.EIP1559 {
+				if _, err := u.FetchDynamicPrice(); err != nil {
+					u.logger.Error(err)
+				}
+			} else {
+				if _, err := u.FetchGasPrice(); err != nil {
+					u.logger.Error(err)
+				}
+			}
+		}
+	}
+}
+
+// GetLegacyGas will fetch the cached gas price value.
+func (u *UniversalEstimator) GetLegacyGas(ctx context.Context, _ []byte, gasLimit uint64, maxPrice *assets.Wei, opts ...feetypes.Opt) (gasPrice *assets.Wei, chainSpecificGasLimit uint64, err error) {
+	chainSpecificGasLimit = gasLimit
+	if gasPrice, err = u.getGasPrice(); err != nil {
+		return
+	}
+
+	if gasPrice.Cmp(maxPrice) > 0 {
+		u.logger.Warnf("estimated gas price: %s is greater than the maximum gas price configured: %s, returning the maximum price instead.", gasPrice, maxPrice)
+		return maxPrice, chainSpecificGasLimit, nil
+	}
+	return
+}
+
+// FetchGasPrice will use eth_gasPrice to fetch and cache the latest gas price from the RPC.
+func (u *UniversalEstimator) FetchGasPrice() (*assets.Wei, error) {
+	ctx, cancel := u.stopCh.CtxCancel(evmclient.ContextWithDefaultTimeout())
+	defer cancel()
+
+	gasPrice, err := u.client.SuggestGasPrice(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch gas price: %s", err)
+	}
+
+	promUniversalEstimatorGasPrice.WithLabelValues(u.chainID.String()).Set(float64(gasPrice.Int64()))
+
+	gasPriceWei := assets.NewWei(gasPrice)
+
+	u.logger.Debugf("fetched new gas price: %v", gasPriceWei)
+
+	u.gasPriceMu.Lock()
+	defer u.gasPriceMu.Unlock()
+	u.gasPrice = gasPriceWei
+	return u.gasPrice, nil
```

Comment:
do we ever do anything with the gasPrice returned here? If not, maybe we should just return the error and call this something like UpdateGasPrice()

Reply:
For now, both `FetchDynamicPrice` and `FetchGasPrice` returned values are not directly used, but I have them as placeholders because in UE's next version, I want to force a fetch and use the value, even if EIP-1559 is not enabled (and vice versa). Consider the following case: if we enable/disable EIP-1559 in the Node and we don't clear in-flight transactions, UE will throw an error and we'll have to clear the txs and restart. This update will cover this edge case.